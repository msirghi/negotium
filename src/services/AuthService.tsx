import Requests from '../common/requests/request';
import { BASE_API_URL_V1 } from '../common/constants/constants';
import { HttpMethod } from '../common/requests/types';
import ServiceResultFactory from '../common/requests/serviceResultFactory';

const login = (email: string, password: string) => {
  return Requests.restApiCall(
    `${BASE_API_URL_V1}/auth/login`,
    HttpMethod.POST,
    {
      email,
      password,
    }
  )
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const register = (email: string, password: string, name: string) => {
  return Requests.restApiCall(`${BASE_API_URL_V1}/users`, HttpMethod.POST, {
    email,
    password,
    name,
  })
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const getUserInfo = () => {
  return Requests.restApiCallWithBearer(
    `${BASE_API_URL_V1}/users/me`,
    HttpMethod.GET
  )
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const getRefreshedToken = () => {
  const body = {
    refresh_token: localStorage.getItem('rt'),
  };
  return Requests.restApiCallWithBearer(
    `${BASE_API_URL_V1}/auth/refresh-token`,
    HttpMethod.POST,
    body
  )
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const AuthService = {
  login,
  register,
  getUserInfo,
  getRefreshedToken,
};

export default AuthService;
