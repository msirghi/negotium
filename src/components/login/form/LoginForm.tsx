import { Fade, TextField } from '@mui/material';
import { Box } from '@mui/system';
import { LoadingButton } from '@mui/lab';
import { FormEvent, useState } from 'react';
import { If } from '../../common/utilities/if/If';
import AuthService from '../../../services/AuthService';
import { useRouter } from 'next/router';
import { useDispatch } from 'react-redux';
import { setAccountInfo } from '../../../redux/account/accountSlice';
import authorizationStore from '../../../common/requests/authorizationStore';
import { useLoginFormStyles } from './styles';
import Routes from '../../../common/config/routes';
import { useTranslation } from 'next-i18next';
import { TextInputChangeEvent } from '../../../common/constants/types';

export const LoginForm = () => {
  const classes = useLoginFormStyles();

  const [loading, setLoading] = useState(false);
  const [error, setError] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  const router = useRouter();
  const dispatch = useDispatch();
  const { t } = useTranslation('auth');

  const onSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setError('');
    setLoading(true);

    try {
      const loginResponse = await AuthService.login(email, password);
      const { data } = loginResponse;
      localStorage.setItem('rt', data.refresh_token);
      authorizationStore.setAuthToken(data.access_token);
      const accountResponse = await AuthService.getUserInfo();
      dispatch(setAccountInfo(accountResponse.data));
      await router.push(Routes.inbox);
    } catch (e) {
      setError((e as Error).message);
    }

    setLoading(false);
  };

  const handleInputChange = (callback: Function) => {
    return (evt: TextInputChangeEvent) => callback(evt.target.value);
  };

  const handleEmailChange = () => handleInputChange(setEmail);

  const handlePasswordChange = () => handleInputChange(setPassword);

  return (
    <form onSubmit={onSubmit}>
      <If condition={!!error}>
        <Fade in>
          <div data-testid={'login-error'} className={classes.error}>
            {error}
          </div>
        </Fade>
      </If>
      <TextField
        autoComplete={'on'}
        value={email}
        inputProps={{ 'data-testid': 'email-input' }}
        onChange={handleEmailChange()}
        style={{ marginTop: 30 }}
        size={'small'}
        fullWidth
        label={t('common.email')}
      />
      <TextField
        value={password}
        autoComplete={'on'}
        inputProps={{ 'data-testid': 'password-input' }}
        onChange={handlePasswordChange()}
        style={{ marginTop: 30 }}
        size={'small'}
        fullWidth
        type={'password'}
        label={t('common.password')}
      />
      <Box className={classes.buttonContainer}>
        <LoadingButton
          data-testid={'submit-button'}
          disabled={!email || !password}
          loading={loading}
          type={'submit'}
          color={'primary'}
          variant={'contained'}
          fullWidth
        >
          {t('registration.signIn')}
        </LoadingButton>
      </Box>
    </form>
  );
};
