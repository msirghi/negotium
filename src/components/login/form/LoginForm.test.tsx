import { SnackbarProvider } from 'notistack';
import { MockReduxProvider } from '../../../common/tests/TestUtils';
import { LoginForm } from './LoginForm';
import renderer, { act } from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import AuthService from '../../../services/AuthService';

const mockPush = jest.fn();

jest.mock('next/router', () => ({
  useRouter: () => ({
    route: 'inbox',
    query: {},
    push: mockPush,
  }),
}));

describe('LoginForm', () => {
  const getComponent = () => {
    return (
      <SnackbarProvider>
        <MockReduxProvider reduxStore={{}}>
          <LoginForm />
        </MockReduxProvider>
      </SnackbarProvider>
    );
  };

  afterEach(() => {
    jest.clearAllMocks();
  })

  it('should match the snapshot', () => {
    const tree = renderer.create(getComponent());
    expect(tree).toMatchSnapshot();
  });

  it('should have disabled submit button by default', () => {
    const { getByTestId } = render(getComponent());
    expect(getByTestId('submit-button')).toBeDisabled();
  });

  it('should handle form submission', () => {
    AuthService.login = jest.fn(() =>
      Promise.resolve({ data: { access_token: 'token' } })
    ) as any;
    AuthService.getUserInfo = jest.fn(() =>
      Promise.resolve({ data: {} })
    ) as any;
    const renderer = render(getComponent());
    const { getByTestId } = renderer;
    const emailInput = getByTestId('email-input');
    const passwordInput = getByTestId('password-input');
    const submitButton = getByTestId('submit-button');

    act(() => {
      fireEvent.change(emailInput, { target: { value: 'email@email.com' } });
      fireEvent.change(passwordInput, { target: { value: 'pass' } });
      fireEvent.click(submitButton);
    });
    expect(AuthService.login).toBeCalled();
  });

  it('should handle the api error', () => {
    AuthService.login = jest.fn(() => Promise.reject('Reject')) as any;
    const {getByTestId} = render(getComponent());
    const emailInput = getByTestId('email-input');
    const passwordInput = getByTestId('password-input');
    const submitButton = getByTestId('submit-button');

    act(() => {
      fireEvent.change(emailInput, { target: { value: 'email@email.com' } });
      fireEvent.change(passwordInput, { target: { value: 'pass' } });
      fireEvent.click(submitButton);
    });
    expect(mockPush).not.toBeCalled();
  });
});
