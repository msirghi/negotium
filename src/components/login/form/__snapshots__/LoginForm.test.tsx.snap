// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match the snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    style={
      Object {
        "marginTop": 30,
      }
    }
  >
    <label
      className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={false}
    >
      common.email
    </label>
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoComplete="on"
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
        data-testid="email-input"
        disabled={false}
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-dhh0px"
        >
          <span>
            common.email
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    style={
      Object {
        "marginTop": 30,
      }
    }
  >
    <label
      className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={false}
    >
      common.password
    </label>
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoComplete="on"
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
        data-testid="password-input"
        disabled={false}
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="password"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-dhh0px"
        >
          <span>
            common.password
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="makeStyles-buttonContainer-2 MuiBox-root css-0"
  >
    <button
      className="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root Mui-disabled css-utm9vu-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
      data-testid="submit-button"
      disabled={true}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="submit"
    >
      registration.signIn
    </button>
  </div>
</form>
`;
